csrc/id_model/predicate_indexing.h

```diff
@@ -35,4 +35,70 @@ std::unordered_map<Val*, Val*> getPredicateIndexReplacementMap(
     bool is_start_predicate,
     ForLoop* unswitched_loop = nullptr);
 
+// Check if a given ExprGroup is a split that needs an additional
+// predicate due to its non-divisibility.
+inline bool isNonDivisibleSplit(const ExprGroup& expr_group) {
+  if (!expr_group->front()->isA<Split>()) {
+    return false;
+  }
+
+  const auto& non_divisible_split_info =
+      GpuLower::current()->nonDivisibleSplitInfo();
+
+  std::vector<PredicateDomainInfo> pred_info_vec;
+
+  // The splitsToPredicate map is for each tensor. Here, it's assumed
+  // that if any tensor has a non-divisible split that's mapped with
+  // expr_group, it should be considered a non-divisible split. This
+  // may result in duplicate predicates, which should be removed by
+  // the expression simplifier.
+  //
+  // For example, suppose tv0 as a 1D tensor of size 16:
+  //
+  // auto tv1 = reshape(tv0, {16}, {2, 8});
+  // tv1->split(1, 3);
+  //
+  // propagate_transformation(to: tv0, from: tv1)
+  //
+  // Here, the split by 3 of tv1 is not included in its non-divisible
+  // split list even though it is indeed non-divisible. The reason is
+  // that the input to the non-divisible split, the inner logical
+  // domain of extent 8, is predicated anyway since it's a logical
+  // domain. Specifically, its predicate should consist of:
+  //
+  // - Predicate for the outer logical domain
+  // - Predicate for the inner logical domain
+  //
+  // However, for tv0, it is indeed included in the non-divisible
+  // split list since the domain of extent 8 is not part
+  // of its logical domain.
+  //
+  // - Predicate for the sole logical domain
+  // - Predicate for the non-divisible split
+  //
+  // This would mean that when generating a predicate for tv1, since
+  // the below check would find a mapping with the non-divisible split
+  // for tv0, the tv1 predicate would be:
+  //
+  // - Predicate for the outer logical domain
+  // - Predicate for the inner logical domain
+  // - Predicate for the non-divisible split
+  //
+  // Here, the last two predicates are redundant since both of them
+  // guard the index with respect to the domain of extent 8, which is
+  // redundant. This is a bit annonying but should have no actual
+  // impact as the redundancy should be removed by the expression
+  // simplifier.
```

Comment:
For `t0`, since its logical domain is the one before the reshape, we would generate:

```
i2 + i1 * 3 + i0 * (ceilDiv(8, 3) * 3) < tv0.logical_shape[0]
```
and because of the ceilDiv, we would also generate:
```
i2 + i1 * 3 < 8
```

The latter is the non-divisible predicate for `t0`. 

For `t1`, since its logical domain is `[2, 8]`, we would generate:

```
i0 < 2
i2 + i1 * 3 < 8
```
And due to the ceilDiv, we would also generate:
```
i2 + i1 * 3 < 8
```

This is also interesting because reshape is always divisible, we can predicate `[2, 8]` for `t0` too. That would probably help index hoisting.